lib/hammer_cli_foreman/api/authenticator.rb

```diff
@@ -0,0 +1,96 @@
+        elsif auth_type == AUTH_TYPES[:oauth_without_2fa]
+          oauth_without_2fa
+        elsif auth_type == AUTH_TYPES[:oauth_with_2fa]
+          ouath_with_2fa
+        end
+      end
+
+      private
+
+      def void_auth
+        VoidAuth.new(_('Using certificate authentication.'))
+      end
+
+      def basic_auth
+        if HammerCLIForeman::Sessions.enabled?
+          authenticator = InteractiveBasicAuth.new(
+            settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'],
+            settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
+          )
+          SessionAuthenticatorWrapper.new(authenticator, uri, auth_type)
+        else
+          username = settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'] || settings.get(:foreman, :username)
+          password = settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
+          if password.nil? && (username == settings.get(:foreman, :username))
+            password = settings.get(:foreman, :password)
+          end
+          InteractiveBasicAuth.new(username, password)
+        end
+      end
+
+      def oauth_without_2fa
+        if HammerCLIForeman::Sessions.enabled?
+          authenticator = Oauth::InteractiveKeycloakAuth.new(
+            settings.get(:_params, :url) || ENV['KEYCLOAK_URL'],
+            settings.get(:_params, :realm) || ENV['KEYCLOAK_REALM'],
+            settings.get(:_params, :client_id) || ENV['KEYCLOAK_CLIENT_ID'],
+            settings.get(:_params, :username) || ENV['KEYCLOAK_USERNAME'],
+            settings.get(:_params, :password) || ENV['KEYCLOAK_PASSWORD']
+          )
+          SessionAuthenticatorWrapper.new(authenticator, uri, auth_type)
+        else
+          url = settings.get(:_params, :url) || ENV['KEYCLOAK_URL'] || settings.get(:foreman, :url)
+          realm = settings.get(:_params, :realm) || ENV['KEYCLOAK_REALM'] || settings.get(:foreman, :realm)
+          client_id = settings.get(:_params, :client_id) || ENV['KEYCLOAK_CLIENT_ID'] || settings.get(:foreman, :client_id)
+          username = settings.get(:_params, :username) || ENV['KEYCLOAK_USERNAME'] || settings.get(:foreman, :username)
+          password = settings.get(:_params, :password) || ENV['KEYCLOAK_PASSWORD']
+          if password.nil? && (username == settings.get(:foreman, :username))
+            password = settings.get(:foreman, :password)
+          end
+          Oauth::InteractiveKeycloakAuth.new(url, realm, client_id, username, password)
+        end
+      end
+
+      def ouath_with_2fa
+        if HammerCLIForeman::Sessions.enabled?
+          authenticator = Oauth::InteractiveKeycloakAuthVia2fa.new(
+            settings.get(:_params, :url) || ENV['KEYCLOAK_URL'],
+            settings.get(:_params, :realm) || ENV['KEYCLOAK_REALM'],
+            settings.get(:_params, :client_id) || ENV['KEYCLOAK_CLIENT_ID']
+          )
+          SessionAuthenticatorWrapper.new(authenticator, uri, auth_type)
+        else
```

Comment:
Same dead code.

Also, `settings.get(:_params, :url/realm/etc)` will not work and don't make sense to have it because `_params` will contain options for `hammer` (main) command parsed from user's input in console/terminal. Although `settings.get(:_params, :username/password)` will work, see https://github.com/theforeman/hammer-cli/blob/8ba92b7375dc31934698d9db0d1f3399037a0011/bin/hammer#L80.